File: python/taflow/ewm_sum.py
"""Persistent exponentially weighted moving sum."""

from typing import Any

import numpy as np

from ._native import EwmSumOperator as _Native
from ._series import as_float64_series


class ExponentiallyWeightedSum:
    """Compute a causal exponentially weighted sum using span ``timeperiod``.

    The smoothing factor is ``2 / (timeperiod + 1)`` and the recurrence is
    ``sum_t = x_t + (1 - alpha) * sum_(t-1)``.
    """

    def __init__(self, timeperiod: int, _input: Any | None = None) -> None:
        self._state = _Native(timeperiod)
        if _input is not None:
            self.extend(_input)

    def append(self, _input: float) -> "ExponentiallyWeightedSum":
        """Append one scalar observation."""
        self._state.append(float(_input))
        return self

    def extend(self, _input: Any) -> "ExponentiallyWeightedSum":
        """Append an aligned input series."""
        self._state.extend(as_float64_series(_input))
        return self

    def compute(self) -> np.ndarray:
        """Return the complete aligned weighted-sum history."""
        return self._state.compute()

    @property
    def value(self) -> float | None:
        """Return the latest weighted sum."""
        return self._state.value

    def reset(self) -> "ExponentiallyWeightedSum":
        """Reset the recurrence and clear output history."""
        self._state.reset()
        return self

File: crates/taflow-core/src/stream/ewm_sum.rs
//! Causal exponentially weighted moving sum.

use crate::error::TaResult;

use super::operator_states::ewm_alpha;

/// Compute an aligned exponentially weighted sum using span=`timeperiod`.
pub fn ewm_sum(input: &[f64], timeperiod: usize) -> TaResult<Vec<f64>> {
    let mut state = ExponentiallyWeightedSum::new(timeperiod)?;
    Ok(input.iter().map(|&value| state.append(value)).collect())
}

/// Persistent exponentially weighted sum with recurrence
/// `sum_t = x_t + (1 - alpha) * sum_(t-1)`.
#[derive(Debug, Clone)]
pub struct ExponentiallyWeightedSum {
    decay: f64,
    value: Option<f64>,
}

impl ExponentiallyWeightedSum {
    /// Create a state using `alpha = 2 / (timeperiod + 1)`.
    pub fn new(timeperiod: usize) -> TaResult<Self> {
        Ok(Self {
            decay: 1.0 - ewm_alpha(timeperiod)?,
            value: None,
        })
    }

    /// Append one observation and return the updated weighted sum.
    pub fn append(&mut self, input: f64) -> f64 {
        let value = input + self.decay * self.value.unwrap_or(0.0);
        self.value = Some(value);
        value
    }

    /// Return the latest weighted sum.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Clear accumulated weight while retaining the configured decay.
    pub fn reset(&mut self) {
        self.value = None;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn matches_polars_documented_example_and_reset() {
        let expected = vec![1.0, 2.5, 4.25];
        assert_eq!(ewm_sum(&[1.0, 2.0, 3.0], 3).unwrap(), expected);
        let mut state = ExponentiallyWeightedSum::new(3).unwrap();
        assert_eq!(state.append(1.0), 1.0);
        assert_eq!(state.append(2.0), 2.5);
        state.reset();
        assert_eq!(state.value(), None);
        assert_eq!(state.append(3.0), 3.0);
    }

    #[test]
    fn rejects_zero_period() {
        assert!(ExponentiallyWeightedSum::new(0).is_err());
    }
}
